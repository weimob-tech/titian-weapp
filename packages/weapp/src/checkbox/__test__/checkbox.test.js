const simulate = require('miniprogram-simulate');
const { getComponent } = require('../../common/test/getComponent');

describe('test component', () => {
  const { getSystemInfoSync } = global.wx;
  const VALUE = 'value';
  beforeAll(() => {
    global.wx.getSystemInfoSync = () => ({
      ...getSystemInfoSync(),
      SDKVersion: '2.10.0'
    });
  });

  afterAll(() => {
    global.wx.getSystemInfoSync = getSystemInfoSync;
  });

  it('should render success', () => {
    const comp = getComponent('checkbox');
    expect(comp).toMatchSnapshot();
  });

  it('should render label, icon, color, size, shape props success', async () => {
    const LABEL = 'label';
    const ICON = 'icon';

    const comp = getComponent('checkbox', {
      label: LABEL,
      icon: ICON,
      color: 'red',
      size: '64',
      shape: '0'
    });
    comp.triggerLifeTime('attached');

    await simulate.sleep(40);

    const expectProps = (selector, expectValue) => {
      expect(comp.querySelector(selector).dom.innerHTML).toEqual(expectValue);
    };

    expectProps('.titian-checkbox-label', LABEL);

    expect(comp).toMatchSnapshot();
  });

  it('should checked do success', (done) => {
    const mockFn = jest.fn();
    const comp = getComponent('checkbox', {
      checked: true
    });

    comp.triggerLifeTime('attached');
    expect(comp).toMatchSnapshot();

    comp.addEventListener('change', mockFn);

    Promise.resolve()
      .then(() => {
        comp.querySelector('.titian-checkbox-icon-wrap').dispatchEvent('tap');
      })
      .then(() => {
        expect(mockFn).toBeCalled();
        comp.querySelector('.titian-checkbox-label').dispatchEvent('tap');
      })
      .then(() => {
        expect(mockFn).toBeCalled();
        done();
      });
  });

  it('should defaultChecked do success', (done) => {
    let defaultChecked = true;
    const mockFn = jest.fn((e) => {
      defaultChecked = !defaultChecked;
      expect(e.detail).toBe(defaultChecked);
    });

    const comp = getComponent('checkbox', {
      defaultChecked,
      value: VALUE
    });

    comp.triggerLifeTime('attached');
    comp.addEventListener('change', mockFn);

    Promise.resolve()
      .then(async () => {
        await simulate.sleep(40);
        comp.querySelector('.titian-checkbox-icon-wrap').dispatchEvent('tap');
        await simulate.sleep(40);
      })
      .then(async () => {
        expect(mockFn).toBeCalledTimes(1);
        comp.querySelector('.titian-checkbox-label').dispatchEvent('tap');
        await simulate.sleep(40);
      })
      .then(() => {
        expect(mockFn).toBeCalledTimes(2);
        done();
      });
  });

  it('should disabled do success', (done) => {
    const mockFn = jest.fn();
    const comp = getComponent('checkbox', {
      disabled: true
    });

    comp.triggerLifeTime('attached');
    expect(comp).toMatchSnapshot();

    comp.addEventListener('change', mockFn);

    Promise.resolve()
      .then(() => {
        comp.querySelector('.titian-checkbox-icon-wrap').dispatchEvent('tap');
      })
      .then(() => {
        expect(mockFn).not.toBeCalled();
        comp.querySelector('.titian-checkbox-label').dispatchEvent('tap');
      })
      .then(() => {
        expect(mockFn).not.toBeCalled();
        done();
      });
  });

  it('should labelDisabled do success', (done) => {
    const mockFn = jest.fn();
    const comp = getComponent('checkbox', {
      labelDisabled: true
    });

    expect(comp).toMatchSnapshot();

    comp.triggerLifeTime('attached');
    comp.addEventListener('change', mockFn);

    Promise.resolve()
      .then(() => {
        comp.instance.updateDataFromParent();
        comp.querySelector('.titian-checkbox-label').dispatchEvent('tap');
      })
      .then(() => {
        expect(mockFn).not.toBeCalled();
        done();
      });
  });

  it('should trigger parent change method success', (done) => {
    const mockFn = jest.fn((e) => {
      expect(e.detail).toStrictEqual([VALUE]);
    });

    const comp = getComponent('checkbox', {
      value: VALUE
    });

    const parent = getComponent('checkbox-group', {
      max: 1
    });

    parent.triggerLifeTime('attached');
    comp.triggerLifeTime('attached');
    parent.addEventListener('change', mockFn);

    Promise.resolve()
      .then(() => {
        comp.instance.setParent(parent.instance);
        comp.querySelector('.titian-checkbox-icon-wrap').dispatchEvent('tap');
      })
      .then(() => {
        expect(mockFn).toBeCalled();
        done();
      });
  });

  it('should select glt max not success', (done) => {
    const mockFn = jest.fn();
    const comp = getComponent('checkbox', {
      value: VALUE
    });

    const parent = getComponent('checkbox-group', {
      max: 2,
      defaultValue: ['placeholder1', 'placeholder2']
    });

    parent.triggerLifeTime('attached');
    comp.triggerLifeTime('attached');
    parent.addEventListener('handleMax', mockFn);

    Promise.resolve()
      .then(async () => {
        comp.instance.setParent(parent.instance);
        comp.instance.updateDataFromParent();
        await simulate.sleep(40);
        comp.querySelector('.titian-checkbox-icon-wrap').dispatchEvent('tap');
        await simulate.sleep(40);
      })
      .then(() => {
        expect(mockFn).toBeCalled();
        done();
      });
  });

  it('should control checked by parent value success', (done) => {
    const mockFn = jest.fn();

    const comp = getComponent('checkbox', {
      value: VALUE
    });

    const parent = getComponent('checkbox-group', {
      max: 2,
      value: ['placeholder1', 'placeholder2', VALUE]
    });

    parent.triggerLifeTime('attached');
    comp.triggerLifeTime('attached');
    parent.addEventListener('change', mockFn);
    Promise.resolve()
      .then(async () => {
        comp.instance.setParent(parent.instance);
        comp.instance.updateDataFromParent();
        await simulate.sleep(40);
        comp.querySelector('.titian-checkbox-icon-wrap').dispatchEvent('tap');
        await simulate.sleep(40);
      })
      .then(() => {
        expect(mockFn).toBeCalled();
        done();
      });
  });
});
