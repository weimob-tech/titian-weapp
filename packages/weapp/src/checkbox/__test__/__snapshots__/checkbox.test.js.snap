// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test component should checked do success 1`] = `
<main>
  <wx-view
    class="ext-class titian-checkbox-checked titian-checkbox-horizontal titian-checkbox"
    style=""
  >
    <wx-view
      class="titian-checkbox-icon-wrap-checked titian-checkbox-icon-wrap"
      style="--base-radius-size:999px; height: 32rpx"
      bind:tap="onChange"
    >
      <ti-icon
        extClass="icon-class titian-checkbox-icon-zoom-in titian-checkbox-icon-checked titian-checkbox-icon-animation titian-checkbox-icon"
      >
        <wx-view
          class="titian-icon-base titian-icon titian-icon-checkbox-hollow ext-class"
          style="line-height:32rpx;height:32rpx;width:32rpx;font-size:32rpx"
          bind:tap="onClick"
        />
      </ti-icon>
    </wx-view>
    <wx-view
      class="titian-checkbox-label"
      bind:tap="onClick"
    />
  </wx-view>
</main>
`;

exports[`test component should disabled do success 1`] = `
<main>
  <wx-view
    class="ext-class titian-checkbox-disabled titian-checkbox-horizontal titian-checkbox"
    style=""
  >
    <wx-view
      class="titian-checkbox-icon-wrap"
      style="--base-radius-size:999px; height: 32rpx"
      bind:tap="onChange"
    >
      <ti-icon
        extClass="icon-class titian-checkbox-icon-animation titian-checkbox-icon"
      >
        <wx-view
          class="titian-icon-base titian-icon titian-icon-checkbox-hollow ext-class"
          style="line-height:32rpx;height:32rpx;width:32rpx;font-size:32rpx"
          bind:tap="onClick"
        />
      </ti-icon>
    </wx-view>
    <wx-view
      class="titian-checkbox-label"
      bind:tap="onClick"
    />
  </wx-view>
</main>
`;

exports[`test component should labelDisabled do success 1`] = `
<main>
  <wx-view
    class="ext-class titian-checkbox-horizontal titian-checkbox"
    style=""
  >
    <wx-view
      class="titian-checkbox-icon-wrap"
      style="--base-radius-size:999px; height: 32rpx"
      bind:tap="onChange"
    >
      <ti-icon
        extClass="icon-class titian-checkbox-icon-animation titian-checkbox-icon"
      >
        <wx-view
          class="titian-icon-base titian-icon titian-icon-checkbox-hollow ext-class"
          style="line-height:32rpx;height:32rpx;width:32rpx;font-size:32rpx"
          bind:tap="onClick"
        />
      </ti-icon>
    </wx-view>
    <wx-view
      class="titian-checkbox-label"
      bind:tap="onClick"
    />
  </wx-view>
</main>
`;

exports[`test component should render label, icon, color, size, shape props success 1`] = `
<main>
  <wx-view
    class="ext-class titian-checkbox-horizontal titian-checkbox"
    style=""
  >
    <wx-view
      class="titian-checkbox-icon-wrap"
      style="height:64rpx;--base-radius-size:0rpx"
      bind:tap="onChange"
    >
      <ti-icon
        extClass="icon-class titian-checkbox-icon-zoom-out titian-checkbox-icon-animation titian-checkbox-icon"
      >
        <wx-view
          class="titian-icon-base titian-icon titian-icon-icon ext-class"
          style="background-color:red;line-height:64rpx;height:64rpx;width:64rpx;font-size:64rpx"
          bind:tap="onClick"
        />
      </ti-icon>
    </wx-view>
    <wx-view
      class="titian-checkbox-label"
      bind:tap="onClick"
    >
      label
    </wx-view>
  </wx-view>
</main>
`;

exports[`test component should render success 1`] = `
<main>
  <wx-view
    class="ext-class titian-checkbox-horizontal titian-checkbox"
    style=""
  >
    <wx-view
      class="titian-checkbox-icon-wrap"
      style="--base-radius-size:999px; height: 32rpx"
      bind:tap="onChange"
    >
      <ti-icon
        extClass="icon-class titian-checkbox-icon-animation titian-checkbox-icon"
      >
        <wx-view
          class="titian-icon-base titian-icon titian-icon-checkbox-hollow ext-class"
          style="line-height:32rpx;height:32rpx;width:32rpx;font-size:32rpx"
          bind:tap="onClick"
        />
      </ti-icon>
    </wx-view>
    <wx-view
      class="titian-checkbox-label"
      bind:tap="onClick"
    />
  </wx-view>
</main>
`;
